Fix digitization bug due to new local fibre ID

cvilelahep opened the pull request:
Propagating the effects of #83 to code related to the digitization.

Since the same mat object is now repeated across all mats in all stations, the fibre IDs are no longer unique. Wherever the code relies on a unique fibre ID the mat and/or station number needs to be added. Conversely, the full paths to the fibre volumes need to be adjusted to use the local fibre numbers.

## Validation
### H6 data example event:
![image](https://user-images.githubusercontent.com/11177614/154680166-52b87957-b68a-4b18-a0ff-33dda2a4fac9.png)
### TI18 background muon simulation example event (zoomed in on SciFi hits, but I checked they match the muon filter pattern):
![image](https://user-images.githubusercontent.com/11177614/154680682-1310743a-68d3-4b5d-8619-ac61ed09467f.png)

ThomasRuf reviewed (commented):
I am now surprised. I thought the previous modification in processhits was made to give ScifiPoint the unique fibre detector ID. But it should not have changed anything on the geometry itself. With the current change, I am afraid this will break getting the correct position for all existing (non MC) data. I am not aware that the Scifi positions for raw data are currently incorrect. I need some time to understand what is happening.

ThomasRuf commented:
Could you test please if these changes are compatible with existing raw data and geo files?
python -i $SNDSW_ROOT/shipLHC/scripts/2dEventDisplay.py -p /eos/experiment/sndlhc/convertedData/commissioning-h6-NewCalib/ -r 283 -g H6/geofile_sndlhc_H6.root

loopEvents(start=0,goodEvents=True,withTrack=2,Setup='H6')

cvilelahep commented:
Yes, I think I'm also seeing some problems with hit positions even in the simulation...
I'll run a few more tests, including raw data, and I'll try to think of a solution.

Even though the scifipoints have a unique ID, some parts of the code (like the sipm-fibre mapping) still rely on the fibres having a unique ID in the geometry, which is what motivates this change.

One possibility is to go back to having multiple mat objects, which will be identical to each other except for the fibre numbering scheme within them. I agree it's not elegant, but it probably has the least impact on the rest of the code.

ThomasRuf commented:
Just to recapitulate: 

Before your changes, when running the simulation, the detectorID was wrongly set for the ScifiPoints. Reading raw data and retrieving positions was ok. 

This issue can be solved with a fix applied only in ProcessHits(). 

An additional modification exist to make the fibre mapping local. What is the advantage of doing this?

cvilelahep commented:
Maybe one much simpler solution is to have M and S always set to `1` in the geometry fibre ID, `1T1RFFF`. In my previous PR I had set M and S to `0`.
I can try checking if this solves the sipm-fibre mapping issue. I assume the sipm-fibre mat is the same for all mats.

ThomasRuf commented:
OK, ProcessHits() fixed and ScifiPoints have correct detectorID.

slowly, I start to understand the problem. It is not related to using sndScifiHits but for making these hits.

A sndScifiHit made by the raw data conversion is OK. The method to get the position of the corresponding SiPM channel works.

Is my understanding of the problem correct?

ThomasRuf commented:
getting the correct fibre position with the old code looks ok:

geo.modules['Scifi'].GetPosition(5036465,A,B)
>>> A.Print()
TVector3 A 3D physics vector (x,y,z)=(-7.342296,57.079608,351.690752) (rho,theta,phi)=(356.368316,9.293389,97.329852)

geo.modules['Scifi'].GetPosition(1036465,A,B)
>>> A.Print()
TVector3 A 3D physics vector (x,y,z)=(-7.236791,57.110400,299.690752) (rho,theta,phi)=(305.169651,10.873404,97.221792)

>>> geo.modules['Scifi'].GetPosition(1016465,A,B)
>>> A.Print()
TVector3 A 3D physics vector (x,y,z)=(-7.236791,31.030401,299.690752) (rho,theta,phi)=(301.379833,6.068884,103.127655)

BUT what looks completely messed up, each Mat seems to contain all the fibres,   
hmat.GetNodes().GetEntries()
42435  while there should be less than 8500 per mat.

Maybe we need indeed a complete cleanup.

cvilelahep commented:
OK, I cleaned up the code a little bit and added some comments to clarify how the "local" fibre IDs are being set.
The fibre volume IDs in the geometry now have the format `1T1RFFF`, i.e., the station and mat numbers are both set to `1`. The ScifiPoint IDs are still as before, with station and mat number fully specified: `STMRFFF`.

I validated with background muon events and H6 data - both look OK to me. Plots in the PR description above.

ThomasRuf commented:
OK. Will also check how the geometry tree looks like now. I assume, I would need to make new geofiles for the different configurations H6/H8 and TI18. What about existing MC data with an existing geofile?

cvilelahep commented:
For the H6 validation I used the existing, old, geofile. This is the command I ran:
```bash
python -i $SNDSW_ROOT/shipLHC/scripts/2dEventDisplay.py -p /eos/experiment/sndlhc/convertedData/commissioning-h6-NewCalib/ -r 283 -g /eos/experiment/sndlhc/convertedData/commissioning-h6-NewCalib/geofile_sndlhc_H6.root
```
For the simulation validation I have been generating events from scratch every time I make a change in the code, so I don't know if it will work with an old geofile -- I'll check.

cvilelahep commented:
I confirmed the code runs on simulated events with an old geofile (generated January 24). In this old file, the scifi hits have the wrong position and are all in the first station, but that's expected because of the ScifiPoint bug.

ThomasRuf reviewed (approved):
OK.
